### PR TITLE
Added fallbacks logic

### PR DESCRIPTION
**Pull request by suriksarkisyan**

### CODE REVIEW

File: Sources/Qonversion/Qonversion/Mappers/QNMapper/QNMapper.m [hunk 1]
```diff
@@ -369,5 +384,15 @@ + (NSInteger)mapInteger:(NSObject *)object orReturn:(NSInteger)defaultValue {
...
 }
 
+- (QONFallbackObject  * _Nullable)mapFallback:(NSDictionary *)data {
+  if (![data isKindOfClass:[NSDictionary class]]) {
+    return nil;
+  }
+  
+  QONFallbackObject *fallback = [QONFallbackObject new];
+  
+  return fallback;
```
Comment (SpertsyanKM) [inline Sources/Qonversion/Qonversion/Mappers/QNMapper/QNMapper.m:394]: Why data is not used?

File: Sources/Qonversion/Qonversion/Mappers/QONFallbacksMapper/QONFallbacksMapper.m [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+//  Created by Suren Sarkisyan on 06.06.2024.
+//  Copyright © 2024 Qonversion Inc. All rights reserved.
+//
+
+#import "QONFallbacksMapper.h"
+#import "QONFallbackObject.h"
+#import "QNMapper.h"
+#import "QONRemoteConfigMapper.h"
+
+@interface QONFallbacksMapper ()
```
Comment (SpertsyanKM) [inline Sources/Qonversion/Qonversion/Mappers/QONFallbacksMapper/QONFallbacksMapper.m:14]: plural form

File: Sources/Qonversion/Qonversion/Mappers/QONFallbacksMapper/QONFallbacksMapper.m [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+{
+  self = [super init];
+  if (self) {
+    _mapper = [QNMapper new];
+    _remoteConfigMapper = [QONRemoteConfigMapper new];
+  }
+  return self;
+}
+
+- (QONFallbackObject  * _Nullable)mapFallback:(NSDictionary *)data {
```
Comment (SpertsyanKM) [inline Sources/Qonversion/Qonversion/Mappers/QONFallbacksMapper/QONFallbacksMapper.m:33]:
```suggestion
- (QONFallbackObject * _Nullable)mapFallbackData:(NSDictionary *)data {
```

File: Sources/Qonversion/Qonversion/Mappers/QONFallbacksMapper/QONFallbacksMapper.m [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  if (![data isKindOfClass:[NSDictionary class]]) {
+    return nil;
+  }
+  
+  QONFallbackObject *fallback = [QONFallbackObject new];
+  
+  fallback.products = [self.mapper mapProducts:data];
+  fallback.offerings = [self.mapper mapOfferings:data];
+  fallback.productsEntitlementsRelation = [self.mapper mapProductsEntitlementsRelationships:data];
```
Comment (SpertsyanKM) [inline Sources/Qonversion/Qonversion/Mappers/QONFallbacksMapper/QONFallbacksMapper.m:42]: If these functions are used only here, I'd offer to rename them to something like `map...FromFallbackFile` so as not to get confused later.

File: Sources/Qonversion/Qonversion/Main/QNProductCenterManager/QNProductCenterManager.m [hunk 5]
```diff
@@ -349,8 +364,8 @@ - (void)purchase:(NSString *)productID offeringID:(NSString *)offeringID complet
     if (self.launchError) {
       __block __weak QNProductCenterManager *weakSelf = self;
       [self launchWithCompletion:^(QONLaunchResult * _Nonnull result, NSError * _Nullable error) {
-        QONLaunchResult *cachedResult = [weakSelf cachedLaunchResult];
-        if (error && !cachedResult) {
+        NSDictionary<NSString *, QONProduct *> *products = [weakSelf getActualProducts];
+        if (error && products.count == 0) {
           run_block_on_main(completion, @{}, error, NO);
```
[the remark this answers is not in the document]
Comment (suriksarkisyan) [reply Sources/Qonversion/Qonversion/Main/QNProductCenterManager/QNProductCenterManager.m:369]: This logic is from the current version of SDK. We can't change it right now

File: Sources/Qonversion/Qonversion/Main/QONRemoteConfigManager/QONRemoteConfigManager.m [hunk 6]
```diff
@@ -99,18 +104,43 @@ - (void)obtainRemoteConfigWithContextKey:(NSString * _Nullable)contextKey comple
     [weakSelf.remoteConfigService loadRemoteConfig:contextKey completion:^(QONRemoteConfig * _Nullable remoteConfig, NSError * _Nullable error) {
       loadingState.isInProgress = NO;
       if (error) {
-        [weakSelf executeRemoteConfigCompletionsWithContextKey:contextKey remoteConfig:nil error:error];
-        completion(nil, error);
-        return;
+        if (error.shouldFireFallback) {
+          weakSelf.fallbackData = weakSelf.fallbackData ?: [weakSelf.fallbacksService obtainFallbackData];
+          QONRemoteConfig *remoteConfig;
+          if (contextKey.length == 0) {
```
[the remark this answers is not in the document]
Comment (suriksarkisyan) [reply Sources/Qonversion/Qonversion/Main/QONRemoteConfigManager/QONRemoteConfigManager.m:110]: Yes it can, but for nil the length is equal to 0

**Review by SpertsyanKM — APPROVED** [review]

File: Sources/Qonversion/Public/Qonversion.h [hunk 7]
```diff
@@ -291,6 +291,12 @@ NS_SWIFT_NAME(remoteConfigList(contextKeys:includeEmptyContextKey:completion:));
  */
 - (void)handlePurchases:(NSArray<QONStoreKit2PurchaseModel *> *)purchasesInfo completion:(nullable QONDefaultCompletionHandler)completion;
 
+/**
+ Call this function to check if fallback file is accessible.
+ @return flag that indicates if fallbacke file is accessible or not.
```
Comment (SpertsyanKM) [inline Sources/Qonversion/Public/Qonversion.h]:
```suggestion
 Call this function to check if the fallback file is accessible.
 @return flag that indicates whether Qonversion was able to read data from the fallback file or not.
```

**Review by SpertsyanKM — APPROVED** [review]

**Pull request merged**